File: examples3d/ccd3.rs
use kiss3d::color::Color;
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

fn create_wall(
    testbed: &mut Testbed,
    world: &mut PhysicsWorld,
    offset: Vector,
    stack_height: usize,
    half_extents: Vector,
) {
    let shift = half_extents * 2.0;
    let mut k = 0;
    for i in 0usize..stack_height {
        for j in i..stack_height {
            let fj = j as f32;
            let fi = i as f32;
            let x = offset.x;
            let y = fi * shift.y + offset.y;
            let z = (fi * shift.z / 2.0) + (fj - fi) * shift.z + offset.z
                - stack_height as f32 * half_extents.z;

            // Build the rigid body.
            let rigid_body = RigidBodyBuilder::dynamic().translation(Vector::new(x, y, z));
            let collider = ColliderBuilder::cuboid(half_extents.x, half_extents.y, half_extents.z);
            let (handle, _) = world.insert(rigid_body, collider);
            k += 1;
            if k % 2 == 0 {
                testbed
                    .set_initial_body_color(handle, Color::new(1., 131. / 255., 244.0 / 255., 1.0));
            } else {
                testbed
                    .set_initial_body_color(handle, Color::new(131. / 255., 1., 244.0 / 255., 1.0));
            }
        }
    }
}

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    let ground_size = 50.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(Vector::new(0.0, -ground_height, 0.0));
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
    let (ground_handle, _) = world.insert(rigid_body, collider);

    /*
     * Create the pyramids.
     */
    let num_z = 8;
    let num_x = 5;
    let shift_y = ground_height + 0.5;
    let shift_z = (num_z as f32 + 2.0) * 2.0;

    for i in 0..num_x {
        let x = i as f32 * 6.0;
        create_wall(
            testbed,
            &mut world,
            Vector::new(x, shift_y, 0.0),
            num_z,
            Vector::new(0.5, 0.5, 1.0),
        );

        create_wall(
            testbed,
            &mut world,
            Vector::new(x, shift_y, shift_z),
            num_z,
            Vector::new(0.5, 0.5, 1.0),
        );
    }

    /*
     * Create two very fast rigid-bodies.
     * The first one has CCD enabled and a sensor collider attached to it.
     * The second one has CCD enabled and a collider attached to it.
     */
    let collider = ColliderBuilder::ball(1.0)
        .density(10.0)
        .sensor(true)
        .active_events(ActiveEvents::COLLISION_EVENTS);
    let rigid_body = RigidBodyBuilder::dynamic()
        .linvel(Vector::new(1000.0, 0.0, 0.0))
        .translation(Vector::new(-20.0, shift_y + 2.0, 0.0))
        .ccd_enabled(true);
    let (sensor_handle, _) = world.insert(rigid_body, collider);

    // Second rigid-body with CCD enabled.
    let collider = ColliderBuilder::ball(1.0).density(10.0);
    let rigid_body = RigidBodyBuilder::dynamic()
        .linvel(Vector::new(1000.0, 0.0, 0.0))
        .translation(Vector::new(-20.0, shift_y + 2.0, shift_z))
        .ccd_enabled(true);
    let (handle, _) = world.insert(rigid_body, collider.clone());
    testbed.set_initial_body_color(handle, Color::new(0.2, 0.2, 1.0, 1.0));

    // Callback that will be executed on the main loop to handle proximities.
    testbed.add_callback(move |mut graphics, physics, events, _| {
        while let Ok(prox) = events.collision_events.try_recv() {
            let color = if prox.started() {
                Color::new(1.0, 1.0, 0.0, 1.0)
            } else {
                Color::new(0.5, 0.5, 1.0, 1.0)
            };

            let parent_handle1 = physics
                .colliders
                .get(prox.collider1())
                .unwrap()
                .parent()
                .unwrap();
            let parent_handle2 = physics
                .colliders
                .get(prox.collider2())
                .unwrap()
                .parent()
                .unwrap();

            if let Some(graphics) = &mut graphics {
                if parent_handle1 != ground_handle && parent_handle1 != sensor_handle {
                    graphics.set_body_color(parent_handle1, color, false);
                }
                if parent_handle2 != ground_handle && parent_handle2 != sensor_handle {
                    graphics.set_body_color(parent_handle2, color, false);
                }
            }
        }
    });

    /*
     * Set up the testbed.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec3::new(100.0, 100.0, 100.0), Vec3::ZERO);
}

File: python/rapier-testbed/rapier_testbed/examples3/ccd3.py
"""Port of ``examples3d/ccd3.rs``.

Two fast balls slam through stacks of pyramids: one ball is a sensor
(triggers ``COLLISION_EVENTS``), the other a regular collider. Both
have CCD enabled so they don't tunnel through the walls.
"""
from __future__ import annotations

import rapier3d as rp

from .._registry import register

CATEGORY = "Dynamics"
NAME = "CCD"


def _create_wall(bodies, colliders, offset, stack_height, half_extents):
    shift_y = half_extents[1] * 2.0
    shift_z = half_extents[2] * 2.0
    for i in range(stack_height):
        for j in range(i, stack_height):
            x = offset[0]
            y = i * shift_y + offset[1]
            z = (
                (i * shift_z / 2.0)
                + (j - i) * shift_z
                + offset[2]
                - stack_height * half_extents[2]
            )
            body = rp.RigidBody.dynamic().translation((x, y, z))
            handle = bodies.insert(body)
            colliders.insert_with_parent(
                rp.Collider.cuboid(half_extents[0], half_extents[1], half_extents[2]),
                handle,
                bodies,
            )


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 50.0
    ground_height = 0.1
    ground_body = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_handle = bodies.insert(ground_body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size),
        ground_handle,
        bodies,
    )

    num_z = 8
    num_x = 5
    shift_y = ground_height + 0.5
    shift_z = (num_z + 2.0) * 2.0

    for i in range(num_x):
        x = i * 6.0
        _create_wall(bodies, colliders, (x, shift_y, 0.0), num_z, (0.5, 0.5, 1.0))
        _create_wall(bodies, colliders, (x, shift_y, shift_z), num_z, (0.5, 0.5, 1.0))

    # Sensor ball with CCD.
    sensor_collider = (
        rp.Collider.ball(1.0)
        .density(10.0)
        .sensor(True)
        .active_events(rp.ActiveEvents.COLLISION_EVENTS)
    )
    sensor_body = (
        rp.RigidBody.dynamic()
        .linvel((1000.0, 0.0, 0.0))
        .translation((-20.0, shift_y + 2.0, 0.0))
        .ccd_enabled(True)
    )
    sensor_handle = bodies.insert(sensor_body)
    colliders.insert_with_parent(sensor_collider, sensor_handle, bodies)

    # Second ball with CCD.
    ball_collider = rp.Collider.ball(1.0).density(10.0)
    ball_body = (
        rp.RigidBody.dynamic()
        .linvel((1000.0, 0.0, 0.0))
        .translation((-20.0, shift_y + 2.0, shift_z))
        .ccd_enabled(True)
    )
    handle = bodies.insert(ball_body)
    colliders.insert_with_parent(ball_collider, handle, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((100.0, 100.0, 100.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
